### PR TITLE
[mlir][vector][memref] Add `alignment` attribute to memory access ops

### PR DESCRIPTION
**Pull request by tyb0807**
Alignment information is important to allow LLVM backends such as AMDGPU
to select wide memory accesses (e.g., dwordx4 or b128). Since this info
is not always inferable, it's better to inform LLVM backends explicitly
about it. Furthermore, alignment is not necessarily a property of the
element type, but of each individual memory access op (we can have
overaligned and underaligned accesses compared to the natural/preferred
alignment of the element type).

This patch introduces `alignment` attribute to memref/vector.load/store
ops.

Follow-up PRs will

1. Propagate the attribute to LLVM/SPIR-V.

2. Introduce `alignment` attribute to other vector memory access ops:
    vector.gather + vector.scatter
    vector.transfer_read + vector.transfer_write
    vector.compressstore + vector.expandload
    vector.maskedload + vector.maskedstore

3. Replace `--convert-vector-to-llvm='use-vector-alignment=1` with a
   simple pass to populate alignment attributes based on the vector
   types.

### CODE REVIEW

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td [hunk 1]
```diff
@@ -1227,7 +1227,45 @@ def LoadOp : MemRef_Op<"load",
...
+                   "Value":$memref,
+                   "ValueRange":$indices,
+                   "bool":$nontemporal), [{
+      return build($_builder, $_state, resultType, memref, indices, nontemporal,
+                   IntegerAttr());
+    }]>,
+    OpBuilder<(ins "TypeRange":$resultTypes,
+                   "Value":$memref,
+                   "ValueRange":$indices,
+                   CArg<"IntegerAttr", "IntegerAttr()">:$alignment), [{
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td]: A lot of these builders could also/instead take a `std::optional<uint32_t>`

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td [hunk 2]
```diff
@@ -1924,13 +1962,30 @@ def MemRef_StoreOp : MemRef_Op<"store",
                        Arg<AnyMemRef, "the reference to store to",
                            [MemWrite]>:$memref,
                        Variadic<Index>:$indices,
-                       DefaultValuedOptionalAttr<BoolAttr, "false">:$nontemporal);
+                       DefaultValuedOptionalAttr<BoolAttr, "false">:$nontemporal,
+                       ConfinedAttr<OptionalAttr<I32Attr>,
+                                    [IntPositive]>:$alignment);
 
   let builders = [
+    OpBuilder<(ins "Value":$valueToStore,
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1973]: Please update documentation - also, we'll want to clarify the units on the alignment (bytes? elements?) and whether there's a power-of-two requirement

File: mlir/unittests/Dialect/MemRef/LoadStoreAlignment.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                                   ValueRange{zero}, alignment);
+  result = mlir::verify(storeOp);
+
+  // Check results
+  EXPECT_TRUE(failed(result));
+  EXPECT_EQ(
+      errorMessage,
+      "'memref.store' op attribute 'alignment' failed to satisfy constraint: "
+      "32-bit signless integer attribute whose value is positive");
+}
```
Comment (krzysz00) [inline mlir/unittests/Dialect/MemRef/LoadStoreAlignment.cpp]: ... this could all be lit tests, you can get rid of this

For one thing `mlir-opt -verify-diagnostics` exists

**Review by krzysz00 — COMMENTED** [review]
Needs more documentation and lit tests, doesn't need the unit tests

I'd be OK with updating the LLVM (and SPIR-V, if applicable) lowerings in this patch or in a stacked-on followup

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td [hunk 4]
```diff
@@ -1227,7 +1227,45 @@ def LoadOp : MemRef_Op<"load",
   let arguments = (ins Arg<AnyMemRef, "the reference to load from",
                            [MemRead]>:$memref,
                        Variadic<Index>:$indices,
-                       DefaultValuedOptionalAttr<BoolAttr, "false">:$nontemporal);
+                       DefaultValuedOptionalAttr<BoolAttr, "false">:$nontemporal,
+                       ConfinedAttr<OptionalAttr<I32Attr>,
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td]: I think the `index` type is more appropriate, since pointers could be aligned to > 4GB in some cases.

**Review by kuhar — CHANGES_REQUESTED** [review]
Thanks for adding this, this will be very useful to have in the IREE codegen.

This also needs llvm/spirv lowering changes and lit tests. We don't need unit tests. See https://mlir.llvm.org/getting_started/TestingGuide/#test-categories

File: mlir/unittests/Dialect/MemRef/LoadStoreAlignment.cpp (hunk 3 above)
Comment (tyb0807) [reply mlir/unittests/Dialect/MemRef/LoadStoreAlignment.cpp]: Yes, I already have all this covered in lit tests above. I just wanted to make sure the new builders work as intended.

**Comment by tyb0807** [comment]
Thanks for the review. Actually, I already have all this covered in lit tests. I just wanted to make sure the new builders work as intended. I guess I can just remove the unit tests?

**Review by matthias-springer — COMMENTED** [review]
Is it possible to build an analysis based on `memref.assume_alignment` instead of adding an alignment attribute to every load/store operation? That's the approach that Triton took (`AxisInfo.cpp`).

**Comment by tyb0807** [comment]
Indeed, but I'm not sure if we can always infer the alignment of load/store op solely from the indexing maths. In case where this is not possible, we would need a way (less automatic) to specify this constraint, right?

**Comment by matthias-springer** [comment]
Can you show an example where that would not work?

**Comment by banach-space** [comment]
Do you ever expect to need something like this (different alignment for different Ops):
```mlir
func.func @test_load_store_alignment(%memref: memref<4xi32>) {
  %c0 = arith.constant 0 : index
  %val = vector.load %memref[%c0] { alignment = 16 } : memref<4xi32>, vector<4xi32>
  vector.store %val, %memref[%c0] { alignment = 32 } : memref<4xi32>, vector<4xi32>
  return
}
```

I am just wondering, why do we need to "decorate" every Op with this attribute? And _what logic_ is meant to take care of it? Why couldn't the alignment be a parameter that's passed to e.g. a conversion pass?

**Comment by matthias-springer** [comment]
Alignment is a property of the memref SSA value. But we don't encode it in the memref type. We have `memref.assume_alignment` as a way to attach alignment information to an SSA value. The alignment can then be queried by a dataflow analysis.

There are two alternatives to this approach:
1. Make the alignment information part of the memref type.
2. Add attributes to each load/store op. (That's what this PR is doing.)

I'd like to make sure that we have a consistent story for dealing with alignment. Having both `memref.assume_alignment` and attributes on various ops seems a bit odd...

**Comment by kuhar** [comment]
> Is it possible to build an analysis based on `memref.assume_alignment` instead of adding an alignment attribute to every load/store operation?

> Alignment is a property of the memref SSA value. But we don't encode it in the memref type.

I don't think this is the case. You can have a memref of `?xi8` that doesn't have any inherent static alignment and the alignment is really a property at each load/store op. You may end up with a memref of bytes as you lower and merge allocations etc. This is also the case with lower level IRs like llvm or spirv, e.g.: https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Memory_Operands.

**Comment by matthias-springer** [comment]
Good point, looks like we may want to have it on the load/store ops.

**Comment by ftynse** [comment]
> You can have a memref of ?xi8 that doesn't have any inherent static alignment and the alignment is really a property at each load/store op. 

At that point, maybe you should use `!ptr.ptr` instead and not a memref. Not necessarily opposing this change, but I don't want to blindly replicate notions from lower-level abstractions like LLVM IR and SPIR-V to a higher-level abstraction.

A stronger argument may be that alignment goes both ways and we can have overaligned and underaligned accesses compared to the natural/preferred alignment of the element type, and those should be reflected somewhere, which is not necessarily a property of the type. Underaligned accesses are more interesting because those may be an optimization hint (aligned accesses are faster) or plainly forbidden by the architecture.

Note that the attribute approach is not precluding a dataflow analysis. We can have an analysis that propagates alignment information to individual operations, e.g., by looking at the structure of subscripts and attributes on previous operations accessing the same value. Attributes can be seen as a way to preserve analysis results.

> Good point, looks like we may want to have it on the load/store ops.

Should we also remove `memref.assume_alignment`? This operation is rather confusing because nothing precludes one from using it repeatedly on the same value and the fact that it is side-effecting (so DCE doesn't remove it) without actually having side effects has been pointed out.

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td [hunk 5]
```diff
@@ -1924,13 +1966,32 @@ def MemRef_StoreOp : MemRef_Op<"store",
...
+                   alignment != 0 ? $_builder.getI64IntegerAttr(alignment) :
+                                    nullptr);
+    }]>,
+ // OpBuilder<(ins "Value":$valueToStore,
+ //                "Value":$memref,
+ //                "ValueRange":$indices,
+ //                "uint64_t":$alignment), [{
+ //   return build($_builder, $_state, valueToStore, memref, indices, false,
+ //                $_builder.getI64IntegerAttr(alignment));
+ // }]>,
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td]:
```suggestion
```

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td [hunk 6]
```diff
@@ -1217,6 +1217,11 @@ def LoadOp : MemRef_Op<"load",
     be reused in the cache. For details, refer to the
     [https://llvm.org/docs/LangRef.html#load-instruction](LLVM load instruction).
 
+    An optional `alignment` attribute allows to specify the byte alignment of the
+    load operation. It must be a positive power of 2. The operation must access
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1220]: Let's have a verifier that checks for it being a power of 2.

Comment (tyb0807) [reply mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1220]: How about a new `IntPowerOf2` `AttrConstraint`?

**Comment by kuhar** [comment]
> At that point, maybe you should use `!ptr.ptr` instead and not a memref.

We want to preserve high-level information so `ptr.ptr` would be too late and won't work for spirv which relies on typed pointers and doesn't lower via `ptr.ptr`.

**Comment by kuhar** [comment]
> Should we also remove `memref.assume_alignment`? This operation is rather confusing because nothing precludes one from using it repeatedly on the same value and the fact that it is side-effecting (so DCE doesn't remove it) without actually having side effects has been pointed out.

I don't have any use for it myself, but I think this still may be useful to denote the alignment of the base pointer if you want to rely on some other analysis or llvm to deduce alignment based on gep offsets from the base pointers. I'd think that if you have multiple `assume_alignment` over the same memref, this should effectively form a conjunction of all the assumptions, e.g.: `and(x div 2, x div 4) => x div 4)` or `and(x div 2, x div 3) ==> x div 6` -- I think this is unsurprising if we only allow pow2 values.

**Comment by krzysz00** [comment]
Two notes:

1. Currently, the lowerings of `vector.load` and `vector.store` to LLVM always assume that the pointer is aligned to the natural alignment of the vector _element_, not the vector. This is behavior users often want to override, either in general or on a case-by-case basis, and having an `alignment` attribute lets people do that
2. I think this'll be a good moment to clarify what `alignment` means for sub-byte types, since we have EmulateNarrowTypes. I'd argue that if the type in the memref has width < 8 bits, then `alignment` should be in units of the number of elements, not the number of bytes. That'll allow the narrow type emulator to use faster lowerings when you, for example, do a `vector.store %v, %m[[...], {alignment = 2 : index} : vector<2xi4>, memref<...xi4>`

File: mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td (hunk 6 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/MemRef/IR/MemRefOps.td:1220]: Is it really worth the complexity? A simple three-line check in the verifier should do the trick.

**Comment by ftynse** [comment]
> We want to preserve high-level information so ptr.ptr would be too late and won't work for spirv which relies on typed pointers and doesn't lower via ptr.ptr.

What high-level information does `memref<?xi8>` have? It's an array of bytes... We need a better interop between pointers and memrefs, in particular taking views of a pointer as a memref in a way that doesn't interfere with aliasing analyses, but I don't see why we should keep abusing memrefs when a pointer is actually needed.

> I don't have any use for it myself, but I think this still may be useful to denote the alignment of the base pointer if you want to rely on some other analysis or llvm to deduce alignment based on gep offsets from the base pointers. I'd think that if you have multiple assume_alignment over the same memref, this should effectively form a conjunction of all the assumptions, e.g.: and(x div 2, x div 4) => x div 4) or and(x div 2, x div 3) ==> x div 6 -- I think this is unsurprising if we only allow pow2 values.

We can certainly take the LCM of two values, that's not the issue here. The issue is that one needs a dataflow analysis to see which assumptions have been stated about which value, e.g., `assume_alignment` may be placed later than a load in a block, but that block may be a body of the loop. And there may be another assumption before and outside that block...

**Comment by ftynse** [comment]
> Currently, the lowerings of vector.load and vector.store to LLVM always assume that the pointer is aligned to the natural alignment of the vector element, not the vector. This is behavior users often want to override, either in general or on a case-by-case basis, and having an alignment attribute lets people do that

There's a lot of logic duplication as well as underdefined cases in this area that I'd love to see cleaned up. For example, we could cast a scalar memref to `memref<... x vector<>>`, at which point it becomes to load with the alignment of vector. Except we don't really have alignment well-specified for multidimensional vectors. And the `vector.load/store` don't preclude non-contiguous leading dimensions...

Again, my concern here is we keep adding alternative ways to do things thus increasing overall complexity of the system. These new ways may be more flexible or general than existing mechanisms, but we _must_ clean those up and ensure some overall coherence.

> I think this'll be a good moment to clarify what alignment means for sub-byte types, since we have EmulateNarrowTypes. I'd argue that if the type in the memref has width < 8 bits, then alignment should be in units of the number of elements, not the number of bytes. That'll allow the narrow type emulator to use faster lowerings when you, for example, do a vector.store %v, %m[[...], {alignment = 2 : index} : vector<2xi4>, memref<...xi4>

Please bring this to discourse. 

IMO, we could say alignment is always specified in the number of bits, not bytes, and be done with it. Having contextual rules to interpret an attribute is a ridiculous amount of unnecessary complexity.

File: mlir/test/Dialect/MemRef/load-store-alignment.mlir [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  %c0 = arith.constant 0 : index
+  %val = memref.load %memref[%c0] { alignment = 16 } : memref<4xi32>
+  memref.store %val, %memref[%c0] { alignment = 16 } : memref<4xi32>
+  return
+}
+
+// -----
+
+func.func @test_invalid_negative_load_alignment(%memref: memref<4xi32>) {
+  // expected-error @+1 {{custom op 'memref.load' 'memref.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
```
Comment (ftynse) [inline mlir/test/Dialect/MemRef/load-store-alignment.mlir]:
```suggestion
  // expected-error @below {{'memref.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
```

Nit: here and below.

File: mlir/test/Dialect/MemRef/load-store-alignment.mlir [hunk 8]
```diff
@@ -0,0 +1,27 @@
+// RUN: mlir-opt -split-input-file -verify-diagnostics %s | FileCheck %s
```
Comment (ftynse) [inline mlir/test/Dialect/MemRef/load-store-alignment.mlir]: Please move the "positive" tests into `test/Dialect/*/ops.mlir` and negative tests into `test/Dialect/*/invalid.mlir`. Both exist in both dialects. We don't want to have a test file per operation.

**Review by ftynse — APPROVED** [review]

**Comment by kuhar** [comment]
> What high-level information does `memref<?xi8>` have? It's an array of bytes...
> why we should keep abusing memrefs when a pointer is actually needed

Going back to the example of merging allocations, we want to do that at the level of memref common for both llvm and spirv. At that point you do have the original memref types and can set the alignment information. Doing the same transformation at the level of spirv and close to llvm (with `ptr.ptr`) seems too late for me, and there's no better place to store alignment info then on load/store ops.

**Review by kuhar — APPROVED** [review]

**Comment by banach-space** [comment]
> Currently, the lowerings of vector.load and vector.store to LLVM always assume that the pointer is aligned to the natural alignment of the vector element, not the vector. This is behavior users often want to override, either in general or on a case-by-case basis, and having an alignment attribute lets people do that

Recently (see https://github.com/llvm/llvm-project/pull/137389), `ConvertVectorToLLVM` was extended to support exactly that kind of override via the `--convert-vector-to-llvm='use-vector-alignment=1' `flag. Isn’t this patch duplicating that logic?

Let’s ask a specific question: could `--convert-vector-to-llvm='use-vector-alignment=1'` be replaced by what’s being added here? If so, great - let’s unify. If not, let’s clarify how they interact.

(Ping @sstamenova for context.)

> Again, my concern here is we keep adding alternative ways to do things thus increasing overall complexity of the system. These new ways may be more flexible or general than existing mechanisms, but we must clean those up and ensure some overall coherence.

+1

> > I think this'll be a good moment to clarify what alignment means for sub-byte types, since we have EmulateNarrowTypes. I'd argue that if the type in the memref has width < 8 bits, then alignment should be in units of the number of elements, not the number of bytes. That'll allow the narrow type emulator to use faster lowerings when you, for example, do a vector.store %v, %m[[...], {alignment = 2 : index} : vector<2xi4>, memref<...xi4>
> 
> Please bring this to discourse.

+1

IMHO, narrow-type emulation is a poor example here. While many patterns in that pass deal with sub-byte types, some also emulate byte-sized elements (see discussion in https://github.com/llvm/llvm-project/issues/131529). That logic has grown organically and lacks cohesion - it’s not an ideal foundation for defining alignment semantics.

---

To be clear, I’m not looking to block this patch - but we now have multiple mechanisms to control alignment:

* The new attributes on `vector.load` / `vector.store` (this patch)
* The `--convert-vector-to-llvm='use-vector-alignment=1'` flag
* `memref.assume_alignment`, which introduces its own complications (see [#144809](https://github.com/llvm/llvm-project/pull/144809), [#144825](https://github.com/llvm/llvm-project/issues/144825))

At the very least, this patch should clarify the relationship between the new attribute and the vector alignment flag in the lowering pass. Otherwise, we risk further fragmentation of behaviour and assumptions around alignment.

File: mlir/test/Dialect/Vector/load-store-alignment.mlir [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  return
+}
+
+// -----
+
+func.func @test_invalid_non_power_of_2_store_alignment(%memref: memref<4xi32>, %val: vector<4xi32>) {
+  // expected-error @+1 {{custom op 'vector.store' 'vector.store' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
+  vector.store %val, %memref[%c0] { alignment = 1 } : memref<4xi32>, vector<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/load-store-alignment.mlir]: This sort of "invalid usage" tests belong in invalid.mlir.

File: mlir/test/Dialect/Vector/load-store-alignment.mlir [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+// CHECK-LABEL: func @test_load_store_alignment
+// CHECK: vector.load {{.*}} {alignment = 16 : i64}
+// CHECK: vector.store {{.*}} {alignment = 16 : i64}
+func.func @test_load_store_alignment(%memref: memref<4xi32>) {
+  %c0 = arith.constant 0 : index
+  %val = vector.load %memref[%c0] { alignment = 16 } : memref<4xi32>, vector<4xi32>
+  vector.store %val, %memref[%c0] { alignment = 16 } : memref<4xi32>, vector<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/load-store-alignment.mlir]: Move to https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/ops.mlir.

**Comment by kuhar** [comment]
> Recently (see #137389), `ConvertVectorToLLVM` was extended to support exactly that kind of override via the `--convert-vector-to-llvm='use-vector-alignment=1' `flag. Isn’t this patch duplicating that logic?

Thanks for linking this, I was not aware of this PR. My assessment is that, in isolation, `use-vector-alignment=1` is closer to being hack that in some cases could have the same effect `alignment =` attributes. Alignment is not inherently a property of the memref type itself, but rather a property of load/store ops, and usually this information is known by the frontend / high-level transformations, so setting it during conversion is too late. For example, in IREE we have a pass to pad shared memory allocations to reduce bank conflicts. This expands allocations by adding extra columns but does not change the types of memory accesses. Using `use-vector-alignment=1` would lead to miscompiles.

The `alignment` attribute is general enough to subsume `use-vector-alignment` -- we can write a simple pass to populate alignment attributes based on the vector types. So in this sense, we may be able to reduce feature duplication in the future.

**Comment by krzysz00** [comment]
Yeah, I remember when landing `--convert-vector-to-llvm='use-vector-alignment=1` I had the sense that it was a temporary solution to work around the fact that we couldn't put `alignment` on vector loads/stores.

I'd argue for removing it in favor of some flavor of `vector-declare-natural-alignment` transform over the vector dialect to reduce that same redundancy.

`memref.assume_alignment` is a rather weird op that, as far as I can tell, exists to allow backends that try to reason about pointer alignment to not stumble over the mysterious pointer-out-of-nowhere that can exist inside memrefs's base.

**Comment by banach-space** [comment]
> The alignment attribute is general enough to subsume use-vector-alignment -- we can write a simple pass to populate alignment attributes based on the vector types. So in this sense, we may be able to reduce feature duplication in the future.

> I'd argue for removing it in favor of some flavor of vector-declare-natural-alignment transform over the vector dialect to reduce that same redundancy.

+1 Lets make sure that there is a clear TODO to that end.

Btw, there is more Ops that access memory:
* `vector.gather` + `vector.scatter`
* `vector.transfer_read` + `vector.transfer_write`
* `vector.compressstore` + `vector.expandload`

What about these?

**Comment by krzysz00** [comment]
Yeah, we probably should get them all - and probably hit them all with nontemporal too while we're here

**Comment by electriclilies** [comment]
Hi! Author of https://github.com/llvm/llvm-project/pull/137389 here :)

Yes, it was a workaround-- we were getting perf issues from vectors being aligned to scalar alignments. I think an `alignment` attribute is a better long term solution, and I'm all for getting rid of the `--convert-vector-to-llvm='use-vector-alignment=1` as long as we're still able to set vector alignments properly. Also, we do have use cases for setting the alignment higher up in the stack, for example during memory planning.

One concern I have with using memref ops for this is that they exist in so many different places in the stack. I could see this becoming messy-- let's say I have a memory planner which tries to pick some alignments, then everything gets lowered into LLVMIR, and then LLVM passes also try to set alignment. It's not clear to me if we should let lower level passes "override" alignment set by higher level passes, or vice versa. And this gets complicated as we interleave default llvm passes with custom ones. It might be worthwhile to expose an interface to let users override how default passes set alignment without making an upstream change. Just something to think about as you start implementing logic to set the alignments.

**Review by electriclilies — APPROVED** [review]

File: mlir/test/Dialect/Vector/load-store-alignment.mlir [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+// -----
+
+func.func @test_invalid_negative_load_alignment(%memref: memref<4xi32>) {
+  %c0 = arith.constant 0 : index
+  // expected-error @+1 {{custom op 'vector.load' 'vector.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
```
Comment (dcaballe) [inline mlir/test/Dialect/Vector/load-store-alignment.mlir]: dup: 'vector.load' 'vector.load'

Comment (krzysz00) [reply mlir/test/Dialect/Vector/load-store-alignment.mlir]: That and it should be in test/Dialect/Vector/invalid.mlir or the like

Comment (krzysz00) [reply mlir/test/Dialect/Vector/load-store-alignment.mlir]: Probably

**Review by krzysz00 — APPROVED** [review]
LGTM aside from the minor test relocation nits

**Comment by banach-space** [comment]
> Yeah, we probably should get them all

Lets do this here, to avoid a split state where some Ops support it while others don't.

**Comment by kuhar** [comment]
> > Yeah, we probably should get them all
> 
> Lets do this here, to avoid a split state where some Ops support it while others don't.

I'd be -1 on this -- I think this PR is already self contained, and similar to the `nontemportal` support, we can extend higher-level ops separately. Smaller PRs are easier to review, and landing this bottom-up makes sense to me.

**Comment by banach-space** [comment]
> I'd be -1 on this -- I think this PR is already self contained, and similar to the nontemportal support, we can extend higher-level ops separately. 

I see this differently.

This is a non-blocker for me, but please make sure that summary provides a justification for not updating all the Ops. Lets also document the next steps for `--convert-vector-to-llvm='use-vector-alignment=1` in the summary. Thanks!

File: mlir/test/Dialect/MemRef/invalid.mlir [hunk 12]
```diff
@@ -1139,3 +1139,21 @@ func.func @expand_shape_invalid_output_shape(
...
+}
+
+// -----
+
+func.func @test_invalid_non_power_of_2_store_alignment(%memref: memref<4xi32>, %val: i32) {
+  %c0 = arith.constant 0 : index
+  // expected-error @below {{'memref.store' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
+  memref.store %val, %memref[%c0] { alignment = 3 } : memref<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/MemRef/invalid.mlir]: Please make sure that these are clustered with other tests for `memref.store` + `memref.load`: https://github.com/llvm/llvm-project/blob/541f33e0751d60b33e75efe0cd436396f27b91ca/mlir/test/Dialect/MemRef/invalid.mlir#L936-L961

File: mlir/test/Dialect/MemRef/ops.mlir [hunk 13]
```diff
@@ -613,3 +613,15 @@ func.func @memref_transpose_map(%src : memref<?x?xf32>) -> memref<?x?xf32, affin
...
+
+// CHECK-LABEL: func @test_load_store_alignment
+// CHECK: memref.load {{.*}} {alignment = 16 : i64}
+// CHECK: memref.store {{.*}} {alignment = 16 : i64}
+func.func @test_load_store_alignment(%memref: memref<4xi32>) {
+  %c0 = arith.constant 0 : index
+  %val = memref.load %memref[%c0] { alignment = 16 } : memref<4xi32>
+  memref.store %val, %memref[%c0] { alignment = 16 } : memref<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/MemRef/ops.mlir]: 1. Try to move these near other tests for `memref.load`/`memref.store` (sadly, these are not really grouped atm, but no harm in trying to improve that)
2. Remove `test` from _test_ function name - that doesn't add any useful context. More on Test naming: https://mlir.llvm.org/getting_started/TestingGuide/#example-improving-test-readability--naming

File: mlir/test/Dialect/Vector/invalid.mlir [hunk 14]
```diff
@@ -2005,3 +2005,21 @@ func.func @vector_store(%dest : memref<?xi8>, %vec : vector<16x16xi8>) {
...
 }
+
+// -----
+
+func.func @test_invalid_negative_load_alignment(%memref: memref<4xi32>) {
+  %c0 = arith.constant 0 : index
+  // expected-error @below {{'vector.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
+  %val = vector.load %memref[%c0] { alignment = -1 } : memref<4xi32>, vector<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/invalid.mlir]: Please move this inside the block for `vector.load` Ops: https://github.com/llvm/llvm-project/blob/541f33e0751d60b33e75efe0cd436396f27b91ca/mlir/test/Dialect/Vector/invalid.mlir#L1985-L1994

Thanks!

Also, I suggest removing `test` and `negative` from test function name. All functions in this file are "test" function to exercise some "negative" (i.e. failing) case. Similar comment for `@test_invalid_non_power_of_2_store_alignment`.

File: mlir/test/Dialect/Vector/ops.mlir [hunk 15]
```diff
@@ -1218,3 +1218,13 @@ func.func @step() {
...
+
+// CHECK-LABEL: func @test_load_store_alignment
+func.func @test_load_store_alignment(%memref: memref<4xi32>) {
+  %c0 = arith.constant 0 : index
+  // CHECK: vector.load {{.*}} {alignment = 16 : i64}
+  %val = vector.load %memref[%c0] { alignment = 16 } : memref<4xi32>, vector<4xi32>
+  // CHECK: vector.store {{.*}} {alignment = 16 : i64}
+  vector.store %val, %memref[%c0] { alignment = 16 } : memref<4xi32>, vector<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/ops.mlir]: Please move this near other tests for `vector.store` + `vector.load`: https://github.com/llvm/llvm-project/blob/541f33e0751d60b33e75efe0cd436396f27b91ca/mlir/test/Dialect/Vector/ops.mlir#L757-L854

Thanks!

**Review by banach-space — COMMENTED** [review]
Thanks for updating the summary, that makes a lot of sense to me 🙏🏻 

> 4. Retire memref.assume_alignment op.

I'm not quite sure whether we have agreement on this? I do support it though and am happy for you to keep it there. 

All in all, LGTM, but I left a few nits to improve our testing "hygiene". Nothing major. I will "approve" once that's addressed.

Thanks for working on this!

File: mlir/test/Dialect/MemRef/invalid.mlir [hunk 16]
```diff
@@ -962,6 +962,24 @@ func.func @test_store_zero_results2(%x: i32, %p: memref<i32>) {
...
+
+// -----
+
+func.func @non_power_of_2_store_alignment(%memref: memref<4xi32>, %val: i32) {
+  %c0 = arith.constant 0 : index
+  // expected-error @below {{'memref.store' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
+  memref.store %val, %memref[%c0] { alignment = 3 } : memref<4xi32>
+  return
+}
```
Comment (banach-space) [inline mlir/test/Dialect/MemRef/invalid.mlir:979]: [nit] Please use consistent naming
```suggestion
func.func @non_power_of_2_load_alignment(%memref: memref<4xi32>) {
  %c0 = arith.constant 0 : index
  // expected-error @below {{'memref.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
  %val = memref.load %memref[%c0] { alignment = -1 } : memref<4xi32>
  return
}

// -----

func.func @non_power_of_2_store_alignment(%memref: memref<4xi32>, %val: i32) {
  %c0 = arith.constant 0 : index
  // expected-error @below {{'memref.store' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
  memref.store %val, %memref[%c0] { alignment = 3 } : memref<4xi32>
  return
}
```

File: mlir/test/Dialect/Vector/invalid.mlir [hunk 17]
```diff
@@ -1995,6 +1995,15 @@ func.func @vector_load(%src : memref<?xi8>) {
...
 
+func.func @negative_load_alignment(%memref: memref<4xi32>) {
+  %c0 = arith.constant 0 : index
+  // expected-error @below {{'vector.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
+  %val = vector.load %memref[%c0] { alignment = -1 } : memref<4xi32>, vector<4xi32>
+  return
+}
+
+// -----
+
```
Comment (banach-space) [inline mlir/test/Dialect/Vector/invalid.mlir:2006]: [nit] Please use consistent naming across the file and within a single PR. You can drop "negative" in this case as all cases in invalid.mlir are "negative" (i.e. they are meant to fail verification). 
```suggestion
func.func @non_power_of_2_load_alignmen(%memref: memref<4xi32>) {
  %c0 = arith.constant 0 : index
  // expected-error @below {{'vector.load' op attribute 'alignment' failed to satisfy constraint: 64-bit signless integer attribute whose value is positive and whose value is a power of two > 0}}
  %val = vector.load %memref[%c0] { alignment = -1 } : memref<4xi32>, vector<4xi32>
  return
}

// -----

```

**Review by banach-space — APPROVED** [review]
Thanks for the updates! I've left a couple of nits, but otherwise LGTM, so approving as is (but please address my comments before merging).

One side note, please note that rebasing and force-pushing is discouraged in LLVM:
* https://llvm.org/docs/GitHub.html#rebasing-pull-requests-and-force-pushes

While rebase is sometimes unavoidable, please avoid squashing commits in a PR - that obfuscates the history and makes reviewing harder. Thanks!

File: mlir/test/Dialect/MemRef/invalid.mlir (hunk 16 above)
Comment (tyb0807) [reply mlir/test/Dialect/MemRef/invalid.mlir:979]: Oh no, `negative` here means < 0, not negative in the sense failing verification

Comment (banach-space) [reply mlir/test/Dialect/MemRef/invalid.mlir:979]: OK, thanks for clarifying. 

This is a bit confusing, we recommend the leading `negative` prefix to label "negative" (i.e. failing) tests:
* https://mlir.llvm.org/getting_started/TestingGuide/#test-naming-convention

How about:
* `@negative_load_alignmen` -> `@invalid_load_alignment`
* `@non_power_of_2_store_alignment` -> `@invalid_store_alignment`

?
